segmentstore/server/src/main/java/io/pravega/service/server/containers/StreamSegmentMapper.java

```diff
@@ -345,63 +345,78 @@ private SegmentState getState(SegmentProperties source, Collection<AttributeUpda
+        SegmentProperties properties = segmentInfo.getProperties();
+        if (properties.isDeleted()) {
             // Stream does not exist. Fail the request with the appropriate exception.
-            failAssignment(streamSegmentInfo.getName(), new StreamSegmentNotExistsException("StreamSegment does not exist."));
-            return FutureHelpers.failedFuture(new StreamSegmentNotExistsException(streamSegmentInfo.getName()));
+            failAssignment(properties.getName(), new StreamSegmentNotExistsException("StreamSegment does not exist."));
+            return FutureHelpers.failedFuture(new StreamSegmentNotExistsException(properties.getName()));
         }
 
-        long streamSegmentId = this.containerMetadata.getStreamSegmentId(streamSegmentInfo.getName(), true);
-        if (isValidStreamSegmentId(streamSegmentId)) {
+        long existingSegmentId = this.containerMetadata.getStreamSegmentId(properties.getName(), true);
+        if (isValidStreamSegmentId(existingSegmentId)) {
             // Looks like someone else beat us to it.
-            completeAssignment(streamSegmentInfo.getName(), streamSegmentId);
-            return CompletableFuture.completedFuture(streamSegmentId);
+            completeAssignment(properties.getName(), existingSegmentId);
+            return CompletableFuture.completedFuture(existingSegmentId);
         } else {
             CompletableFuture<Long> logAddResult;
             StreamSegmentMapping mapping;
             if (isValidStreamSegmentId(parentStreamSegmentId)) {
                 // Transaction.
                 SegmentMetadata parentMetadata = this.containerMetadata.getStreamSegmentMetadata(parentStreamSegmentId);
                 assert parentMetadata != null : "parentMetadata is null";
-                TransactionMapOperation op = new TransactionMapOperation(parentStreamSegmentId, streamSegmentInfo);
-                mapping = op;
+                TransactionMapOperation op = new TransactionMapOperation(parentStreamSegmentId, properties);
+                mapping = transferSegmentId(segmentInfo, op);
                 logAddResult = this.durableLog.add(op, timeout);
             } else {
                 // Standalone StreamSegment.
-                StreamSegmentMapOperation op = new StreamSegmentMapOperation(streamSegmentInfo);
-                mapping = op;
+                StreamSegmentMapOperation op = new StreamSegmentMapOperation(properties);
+                mapping = transferSegmentId(segmentInfo, op);
                 logAddResult = this.durableLog.add(op, timeout);
             }
 
             return logAddResult
-                    .thenApply(seqNo -> completeAssignment(streamSegmentInfo.getName(), mapping.getStreamSegmentId()));
+                    .thenApply(seqNo -> completeAssignment(properties.getName(), mapping.getStreamSegmentId()));
         }
     }
 
     /**
+     * Transfers the segment id from the given segment info to the given mapping, if any,
```

Comment:
Here and in the method name, I found the word "transfer" confusing, as it seemed to apply some sort of ownership, like that some other mapping could lose the segmentId.